lib/client/ipam_block_reader_writer.go

```diff
@@ -309,24 +316,36 @@ func randomBlockGenerator(pool cnet.IPNet) func() *cnet.IPNet {
 	ones, size := pool.Mask.Size()
 	prefixLen := size - ones
 	numBlocks := int(math.Exp2(float64(prefixLen))) / blockSize
-
-	// Generate a randomly ordered slice of block indexes.
-	source := rand.NewSource(time.Now().UnixNano())
-	randm := rand.New(source)
-	idxs := randm.Perm(numBlocks)
 	i := 0
+	if !indexInit {
+		// Start at a random offset index
+		source := rand.NewSource(time.Now().UnixNano())
+		randm := rand.New(source)
+		indexInit = true
+		initialIndex = randm.Intn(numBlocks)
+		i = initialIndex
+	} else {
+		i = currentIndex
+	}
+
 	return func() *cnet.IPNet {
-		// Pop a block index off of the indexes slice.
-		if len(idxs) != 0 {
-			i, idxs = idxs[len(idxs)-1], idxs[:len(idxs)-1]
-		} else {
+		if i < numBlocks {
```

Comment:
I don't think this is the right check - this assumes we're starting at index 0!

We want to set `i = 0` when `!pool.Contains(ip)`, and then return nil when `currentIndex == initialIndex`.